File: 153-find-minimum-in-rotated-sorted-array/153-find-minimum-in-rotated-sorted-array.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        
        if nums[0] <= nums[len(nums)-1]:
            return nums[0]
        
        start, end = 0, len(nums)-1
        
        while start <= end:
            mid = (start + end) // 2
            
            if nums[mid] > nums[mid+1]:
                return nums[mid+1]
            
            elif nums[mid] < nums[mid - 1]:
                return nums[mid]
            
            elif nums[start] <= nums[mid]:
                start = mid + 1
            
            else:
                end = mid - 1
        
        return -1

File: 153-find-minimum-in-rotated-sorted-array/153-find-minimum-in-rotated-sorted-array.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
        
        int lo=0;
        int hi=nums.size()-1;
        
        if(nums[0] <= nums[hi]) return nums[0];
        
        while(lo<=hi){
            
            int mid = lo + (hi-lo)/2;
            
            if(nums[mid] > nums[mid+1]){
                return nums[mid+1];
            }
            
            else if(nums[mid-1] > nums[mid]){
                return nums[mid];
            }
            
            if(nums[lo] <= nums[mid]){
                lo = mid+1;
            }
            else if(nums[mid] <= nums[hi]){
                hi = mid-1;
            }
        }
        
        return -1;
    }
};